sycl/source/detail/stream_impl.cpp

```diff
@@ -26,41 +27,40 @@ stream_impl::stream_impl(size_t BufferSize, size_t MaxStatementSize,
 stream_impl::stream_impl(size_t BufferSize, size_t MaxStatementSize,
                          const property_list &PropList)
     : BufferSize_(BufferSize), MaxStatementSize_(MaxStatementSize),
-      PropList_(PropList) {
-  // We need to store stream buffers in the scheduler because they need to be
-  // alive after submitting the kernel. They cannot be stored in the stream
-  // object because it causes loop dependency between objects and results in
-  // memory leak.
-  // Allocate additional place in the stream buffer for the offset variable and
-  // the end of line symbol.
-  detail::Scheduler::getInstance().allocateStreamBuffers(
-      this, BufferSize + OffsetSize + 1 /* size of the stream buffer */,
-      MaxStatementSize + FLUSH_BUF_OFFSET_SIZE /* size of the flush buffer */);
+      PropList_(PropList), Buf_(range<1>(BufferSize + OffsetSize + 1)),
+      FlushBuf_(range<1>(MaxStatementSize + FLUSH_BUF_OFFSET_SIZE)) {
+  // Additional place is allocated in the stream buffer for the offset variable
+  // and the end of line symbol. Buffers are created without host pointers so
+  // that they are released in a deferred manner. Disable copy back on buffer
+  // destruction. Copy is scheduled as a host task which fires up as soon as
+  // kernel has completed exectuion.
+  Buf_.set_write_back(false);
+  FlushBuf_.set_write_back(false);
+  // Initialize stream buffer with zeros, this is needed for two reasons:
+  // 1. We don't need to care about end of line when printing out
+  // streamed data.
+  // 2. Offset is properly initialized.
+  host_accessor Acc{Buf_};
+  char *Ptr = Acc.get_pointer();
+  std::memset(Ptr, 0, Buf_.size());
 }
 
 // Method to provide an access to the global stream buffer
 GlobalBufAccessorT stream_impl::accessGlobalBuf(handler &CGH) {
-  return detail::Scheduler::getInstance()
-      .StreamBuffersPool.find(this)
-      ->second->Buf.get_access<sycl::access::mode::read_write>(
-          CGH, range<1>(BufferSize_), id<1>(OffsetSize));
+  return Buf_.get_access<sycl::access::mode::read_write>(
+      CGH, range<1>(BufferSize_), id<1>(OffsetSize));
```

Comment:
Can we use `accessor`'s CTAD here?

Reply:
I'm having trouble using these in our library code, getting a CTAD compilation error with gcc (clang works just fine, we don't see any issues in SYCL applications after all). Not sure why, I'll look into it and address the stylistic accessor changes in this file separately, if you don't mind.